src/vm/types.rs

```diff
@@ -480,6 +570,70 @@ impl TypeSignature {
                         list_dimensions: None }
     }
 
+    pub fn new_option(inner_type: TypeSignature) -> TypeSignature {
+        TypeSignature::new_atom(
+            AtomTypeIdentifier::OptionalType(
+                Box::new(inner_type)))
+    }
+
+    fn make_union_response_type(response_type_a: &(TypeSignature, TypeSignature),
```

Comment:
I'm confused by this method.  If I have a `ResponseType<A, B>` and another `ResponseType<C, D>`, this method generates `ResponseType<(A | B), (C | D)>`?  What purpose does this serve?

Reply:
It specifically won't allow that -- it takes `ResponseType<*, A>` and `ResponseType<B, *>` and creates   `ResponseType<B, A>` -- its so that the type checker can integrate a CFG that looks like:

```
  (if bool (ok Type1) (err Type2))
```

into the type `Response<Type1, Type2>`

Reply:
Ah, okay.  This makes a lot more sense now.